src/diff.rs

```diff
@@ -182,91 +201,683 @@ fn diff_attributes(change_list: &mut ChangeListBuilder, old: &[Attribute], new:
+    // Map from each index in `new` to the index of the node in `old` that
+    // has the same key.
+    let mut new_index_to_old_index = Vec::with_capacity(new.len());
+    new_index_to_old_index.extend(new.iter().map(|n| {
+        let key = n.key();
+        if let Some(&i) = old_key_to_old_index.get(&key) {
+            shared_keys.insert(key);
+            i
+        } else {
+            u32::MAX as usize
+        }
+    }));
+
+    // If none of the old keys are reused by the new children, then we
+    // remove all the remaining old children and create the new children
+    // afresh.
+    if shared_suffix_count == 0 && shared_keys.is_empty() {
+        if shared_prefix_count == 0 {
+            remove_all_children(change_list, registry, old);
         } else {
-            debug_assert!(pushed);
             change_list.pop_push_next_sibling();
+            remove_self_and_next_siblings(change_list, registry, &old[shared_prefix_count..]);
         }
+        create_and_append_children(cached_set, change_list, registry, new, cached_roots);
+        return;
+    }
+
+    // The longest increasing subsequence within `new_index_to_old_index` in
+    // reverse order (since `lis_with` adds the indices in reverse
+    // order). We will leave these nodes in place in the DOM, and only move
+    // nodes that are not part of the LIS. This results in the minimum
+    // number of DOM nodes moved.
+    let mut reverse_lis = Vec::with_capacity(new_index_to_old_index.len());
+    let mut predecessors = vec![0; new_index_to_old_index.len()];
+    let mut starts = vec![0; new_index_to_old_index.len()];
+    longest_increasing_subsequence::lis_with(
+        &new_index_to_old_index,
+        &mut reverse_lis,
+        |a, b| a < b,
+        &mut predecessors,
+        &mut starts,
+    );
+
+    // Save each of the old children whose keys are reused in the new
+    // children.
+    let mut old_index_to_temp = vec![u32::MAX; old.len()];
+    let mut start = 0;
+    loop {
+        let end = (start..old.len())
+            .find(|&i| {
+                let key = old[i].key();
+                !shared_keys.contains(&key)
+            })
+            .unwrap_or(old.len());
+
+        if end - start > 0 {
+            let mut t = change_list.save_children_to_temporaries(
+                shared_prefix_count + start,
+                shared_prefix_count + end,
+            );
+            for i in start..end {
+                old_index_to_temp[i] = t;
+                t += 1;
+            }
+        }
+
+        debug_assert!(end <= old.len());
+        if end == old.len() {
+            break;
+        } else {
+            start = end + 1;
+        }
+    }
+
+    // Remove any old children whose keys were not reused in the new
+    // children. Remove from the end first so that we don't mess up indices.
+    let mut removed_count = 0;
+    for (i, old_child) in old.iter().enumerate().rev() {
+        if !shared_keys.contains(&old_child.key()) {
```

Comment:
Hm ok I may again be very confused, but this sort of looks like it's preserving everything with a key of `u32::MAX`?